### ISSUE TITLE
BUG: DataFrame.from_records()'s columns argument doesn't work on Numpy's structured array

### ISSUE THREAD

**Issue by ammar-qazi** [pandas-dev/pandas#59717]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np

data = np.array([
    ('John', 25, 'New York', 50000),
    ('Jane', 30, 'San Francisco', 75000),
    ('Bob', 35, 'Chicago', 65000),
    ('Alice', 28, 'Los Angeles', 60000)
], dtype=[('name', 'U10'), ('age', 'i4'), ('city', 'U15'), ('salary', 'i4')])

df = pd.DataFrame.from_records(data, columns=['salary', 'city', 'age', 'name'])

print(df)
```

### Issue Description

The `columns` parameter doesn't appear to reorder the columns when the input data is a structured array, even though the documentation says it should. (It does do this for other data types.) In addition, the argument doesn't filter the columns like it does for other input types like dict (a behavior which isn't documented yet, but which will be -- see #59670).

### Expected Behavior

According to the discussion in #15319 and #59670, DataFrame.from_records()'s columns argument should allow the users to reorder columns (and include only specific columns) from the Numpy's structured array the way it works for data passed as a dictionary. 

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit              : a671b5a8bf5dd13fb19f0e88edc679bc9e15c673
python              : 3.10.12.final.0
python-bits         : 64
OS                  : Linux
OS-release          : 6.1.85+
Version             : #1 SMP PREEMPT_DYNAMIC Thu Jun 27 21:05:47 UTC 2024
machine             : x86_64
processor           : x86_64
byteorder           : little
LC_ALL              : en_US.UTF-8
LANG                : en_US.UTF-8
LOCALE              : en_US.UTF-8

pandas              : 2.1.4
numpy               : 1.26.4
pytz                : 2024.1
dateutil            : 2.8.2
setuptools          : 71.0.4
pip                 : 24.1.2
Cython              : 3.0.11
pytest              : 7.4.4
hypothesis          : None
sphinx              : 5.0.2
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : 4.9.4
html5lib            : 1.1
pymysql             : None
psycopg2            : 2.9.9
jinja2              : 3.1.4
IPython             : 7.34.0
pandas_datareader   : 0.10.0
bs4                 : 4.12.3
bottleneck          : None
dataframe-api-compat: None
fastparquet         : None
fsspec              : 2024.6.1
gcsfs               : 2024.6.1
matplotlib          : 3.7.1
numba               : 0.60.0
numexpr             : 2.10.1
odfpy               : None
openpyxl            : 3.1.5
pandas_gbq          : 0.23.1
pyarrow             : 14.0.2
pyreadstat          : None
pyxlsb              : None
s3fs                : None
scipy               : 1.13.1
sqlalchemy          : 2.0.32
tables              : 3.8.0
tabulate            : 0.9.0
xarray              : 2024.6.0
xlrd                : 2.0.1
zstandard           : None
tzdata              : 2024.1
qtpy                : None
pyqt5               : None

</details>

**Comment by cjerdonek** [comment]
Since the "filtering" aspect isn't documented yet, it seems like the emphasis in this issue should be on the fact that the columns argument doesn't do any reordering, which _is_ documented. So I would rephrase the issue description to something like this (and provide another example as needed)--

> The Columns parameter doesn't appear to reorder the columns when the input data is a structured array, even though the documentation says it should. (It does do this for other data types.) In addition, the argument doesn't filter the columns like it does for other input types like dict (a behavior which isn't documented yet, but which will be -- see https://github.com/pandas-dev/pandas/issues/59670).

**Comment by ammar-qazi** [comment]
@cjerdonek I've adjusted the issue text to reflect this.

**Comment by cjerdonek** [comment]
Thanks!

**Comment by rhshadrach** [comment]
Thanks for the report, confirmed on main. Further investigations and PRs to fix are welcome!

**Comment by VibavariG** [comment]
Take
